paddle/fluid/operators/controlflow/while_op.cc

```diff
@@ -87,14 +101,48 @@ class WhileOp : public framework::OperatorBase {
     bool is_test = Attr<bool>("is_test");
     auto &skip_vars = Attr<std::vector<std::string>>(kSkipEagerDeletionVars);
     VLOG(2) << GetSkipEagerDeletionVarsDebugString(skip_vars);
-
     auto ctx = executor.Prepare(*program, block->ID(), skip_vars);
     if (!is_test) {
       while (cond_data) {
         auto &current_scope = scope.NewScope();
         step_scopes->push_back(&current_scope);
+        // Each loop the value of LoDTensor in input would be saved sub_scope to
+        // ensure that the correct values can be used in the gradient
+        // calculation.
+        auto input_names = Inputs(kX);
+        for (auto &input_var_name : input_names) {
+          if (std::find(no_copy_var_names.begin(), no_copy_var_names.end(),
+                        input_var_name) == no_copy_var_names.end()) {
+            auto input_var_rename = input_var_name + "@RENAME";
+            auto *ig_outside_var = scope.FindVar(input_var_name);
+            if (ig_outside_var->IsType<framework::LoDTensor>()) {
+              auto ig_outside_var_tensor = ig_outside_var->Get<LoDTensor>();
+              auto *ig_inside_var_tensor =
+                  current_scope.Var(input_var_rename)->GetMutable<LoDTensor>();
+              framework::TensorCopy(ig_outside_var_tensor, dev_place,
+                                    ig_inside_var_tensor);
+              ig_inside_var_tensor->set_lod(ig_outside_var_tensor.lod());
+            }
+          }
+        }
         executor.RunPreparedContext(ctx.get(), &current_scope, false, true,
                                     true);
+        std::vector<std::string> erase_vars;
```

Comment:
You can add those erase_vars when you rename them instead of finding them after executor running, in this case, you don't need to find again and code is shorter.

Reply:
Thanks, done.